### PR TITLE
feat: logical query conditions within one column and $not

### PR DESCRIPTION
**Pull request by JimmyDaddy**
Related to #74

### CODE REVIEW

File: lib/drivers/abstract/spellbook.js [hunk 1]
```diff
@@ -174,7 +173,8 @@ function formatOpExpr(spell, ast) {
   else if ('!~-'.includes(name) && params.length == 1) {
     return `${name} ${params[0]}`;
   }
-  else if (args[1].type == 'literal' && args[1].value == null) {
+  else
+  if (args[1].type == 'literal' && args[1].value == null && !isLogicalOp(ast)) {
```
Comment (JimmyDaddy) [inline lib/drivers/abstract/spellbook.js:177]: fix
```
Post.where({
    title: {
          $or: [
            {
              $ne: 'Leah'
            },
            null
          ],
        }
      })
```

Comment (cyjake) [inline lib/drivers/abstract/spellbook.js:177]: no need to break line here

File: lib/spell.js [hunk 2]
```diff
@@ -63,9 +67,13 @@ function parseConditions(conditions, ...values) {
  * @returns {Object}
  */
 function parseObjectValue(value) {
-  return value instanceof module.exports
-    ? { type: 'subquery', value }
-    : parseExpr('?', value);
+  if (value instanceof module.exports) return { type: 'subquery', value };
+  // value maybe a conditions
```
Comment (cyjake) [inline lib/spell.js:71]: value might be an object condition

File: lib/spell.js [hunk 3]
```diff
@@ -142,13 +151,43 @@ function parseLogicalObjectConditionValue(value) {
...
  * @param {Object|Object[]} value logical operands
  */
 function parseLogicalObjectCondition(name, value) {
   const operator = LOGICAL_OPERATOR_MAP[name];
   const conditions = parseLogicalObjectConditionValue(value);
   const args = conditions.reduce((res, condition) => {
+    if (!isPlainObject(condition)) return res.concat({
+      type: 'literal',
+      value: condition,
+    });
```
Comment (cyjake) [inline lib/spell.js:190]: multi line if/else branches should be wrapped in `{}`

```suggestion
    if (!isPlainObject(condition)) {
      return res.concat({
        type: 'literal',
        value: condition,
      });
    }
```

File: lib/spell.js [hunk 4]
```diff
@@ -166,13 +205,12 @@ function parseLogicalObjectCondition(name, value) {
  * { foo: { $gt: new Date(2012, 4, 15) } }
  * { foo: { $between: [1, 10] } }
  * @param {Object} conditions
- */
+  */
```
Comment (cyjake) [inline lib/spell.js:208]: unnecessary space

File: lib/spell.js [hunk 5]
```diff
@@ -185,12 +223,24 @@ function parseObjectConditions(conditions) {
     }
     else if (isOperatorCondition(value)) {
       result.push(parseOperatorCondition(name, value));
+    } else if(isOperator(name)) {
```
Comment (cyjake) [inline lib/spell.js:226]: space after `if`

File: lib/spell.js [hunk 6]
```diff
@@ -185,12 +223,24 @@ function parseObjectConditions(conditions) {
...
+      // if name is an common operator
```
Comment (cyjake) [inline lib/spell.js:227]: is a common operator

File: lib/spell.js (hunk 5 above)
Comment (cyjake) [reply lib/spell.js:226]: `else if` should be broke into newline, since the convention here is so

File: test/unit/spell.test.js [hunk 7]
```diff
@@ -95,6 +95,237 @@ describe('=> Select', function() {
...
+            $not: [
+              100,
+              { $lte: 2 }
+            ]
+          }
+        }).toSqlString(),
+        "SELECT * FROM `articles` WHERE (`title` = 'Leah' OR `title` != 'Leah') AND (`is_private` >= 1 AND `is_private` <= 6) AND (NOT (`author_id` = 100 AND `author_id` <= 2)) AND `gmt_deleted` IS NULL"
+      );
+    })
+  })
```
Comment (cyjake) [inline test/unit/spell.test.js:327]: 👍 nice

File: lib/drivers/abstract/spellbook.js [hunk 8]
```diff
@@ -196,8 +196,80 @@ function formatOpExpr(spell, ast) {
     } else {
       throw new Error(`Invalid operator ${name} against ${args[1].value}`);
     }
-  }
-  else {
+  } else if (args[1].type == 'op' && Array.isArray(args[1].args) && !isLogicalOp(ast)) {
```
Comment (cyjake) [inline lib/drivers/abstract/spellbook.js:199]:
```suggestion
  }
  // { 
  //   foo: { $or: [ ... ] },
  // }  
  else if (args[1].type == 'op' && Array.isArray(args[1].args) && !isLogicalOp(ast)) {
```

File: lib/spell.js [hunk 9]
```diff
@@ -142,13 +151,43 @@ function parseLogicalObjectConditionValue(value) {
...
+ *       },
+ *     ]
+ *   }
+ * }
+ * {
+ *   title: {
+ *     $not: [
+ *       'Leah',
+ *       'jss'
+ *     ]
```
Comment (cyjake) [inline lib/spell.js:177]: does this get translated into `title NOT IN ('Leah', 'jss')`?

Comment (JimmyDaddy) [reply lib/spell.js:177]: fixed

File: .eslintrc [hunk 10]
```diff
@@ -17,6 +17,7 @@
     "no-undef": 2,
     "no-underscore-dangle": 0,
     "no-use-before-define": [2, "nofunc"],
-    "no-unused-vars": [2, { "vars": "all", "args": "none" }]
+    "no-unused-vars": [2, { "vars": "all", "args": "none" }],
+    "keyword-spacing": "error"
```
Comment (JimmyDaddy) [inline .eslintrc:21]: Enforce consistent spacing after keywords

**Review by cyjake — APPROVED** [review]
+1

**Pull request merged**